### PR TITLE
client/chunkserver/curve-nbd: support 512 aligned IO requests

### PR DESCRIPTION
**Pull request by wu-hanqing**
### What problem does this PR solve?

Issue Number: close #xxx 

Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

Side effects(Breaking backward compatibility? Performance regression?): 

### Check List

- [ ] Relevant documentation/comments is changed or added
- [ ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

File: src/client/splitor.cpp [hunk 1]
```diff
@@ -393,6 +401,54 @@ bool Splitor::MarkDiscardBitmap(IOTracker* iotracker, FileSegment* fileSegment,
     return true;
 }
 
+uint64_t Splitor::ProcessUnalignedRequests(const off_t currentOffset,
```
Comment (SeanHai) [inline src/client/splitor.cpp:404]: If request is unaligned, need artificial alignment?

File: src/common/fast_align.h [hunk 2]
```diff
@@ -0,0 +1,61 @@
+/*
+Copyright 2015 Glen Joseph Fernandes
+(glenjofe@gmail.com)
+Distributed under the Boost Software License, Version 1.0.
+(http://www.boost.org/LICENSE_1_0.txt)
```
Comment (ilixiaocui) [inline src/common/fast_align.h:5]: The corresponding unit test

File: conf/chunkserver.conf.example [hunk 3]
```diff
@@ -15,6 +15,8 @@ global.meta_page_size=4096
 # clone chunk允许的最长location长度
 global.location_limit=3000
 
+global.min_io_alignment=512
```
[the remark this answers is not in the document]
Comment (wu-hanqing) [reply conf/chunkserver.conf.example:18]: fix

File: conf/snap_client.conf [hunk 4]
```diff
@@ -159,3 +159,9 @@ discard.enable=false
 discard.granularity=4096
 # discard cleanup task delay times in millisecond
 discard.taskDelayMs=60000
+
+##### alignment #####
+# default alignment
+global.alignment.commonVolume=512
+# alignment for clone volume
+global.alignment.cloneVolume=4096
```
[the remark this answers is not in the document]
Comment (wu-hanqing) [reply conf/snap_client.conf:167]: added annoation

File: src/chunkserver/datastore/chunkserver_chunkfile.cpp [hunk 5]
```diff
@@ -30,6 +30,8 @@
 namespace curve {
 namespace chunkserver {
 
+uint32_t kMinIOAlignment = 512;
```
[the remark this answers is not in the document]
Comment (wu-hanqing) [reply src/chunkserver/datastore/chunkserver_chunkfile.cpp:33]: changed it to gflags

File: src/client/request_closure.h [hunk 6]
```diff
@@ -149,6 +148,13 @@ class CURVE_CACHELINE_ALIGNMENT RequestClosure
         return suspendRPC_;
     }
 
+ protected:
+    // closure的request信息
```
[the remark this answers is not in the document]
Comment (wu-hanqing) [reply src/client/request_closure.h:152]: fix

File: src/client/splitor.cpp (hunk 1 above)
Comment (wu-hanqing) [reply src/client/splitor.cpp:404]: yes, PaddingReadClosure will form an aligned request.

File: include/chunkserver/chunkserver_common.h [hunk 7]
```diff
@@ -124,7 +124,7 @@ inline std::string ToGroupIdString(const LogicPoolID &logicPoolId,
 #define ToGroupIdStr   ToGroupIdString
 
 // TODO(wudmeiao): 是否需要考虑可配置
-const uint32_t kOpRequestAlignSize = 4096;
+// const uint32_t kOpRequestAlignSize = 512;  // TODO: check this
```
[the remark this answers is not in the document]
Comment (wu-hanqing) [reply include/chunkserver/chunkserver_common.h:127]: missing delete, fix

File: nbd/src/define.h [hunk 8]
```diff
@@ -88,6 +87,8 @@ struct NBDConfig {
     int retry_times = 25;
     // unmap重试之间的睡眠间隔
     int sleep_ms = 200;
+    // device's block size
+    int block_size = 4096;
```
[the remark this answers is not in the document]
Comment (wu-hanqing) [reply nbd/src/define.h:91]: it's just device's block size, but it's default value is 4096

File: robot/curve_robot.txt [hunk 9]
```diff
@@ -1449,11 +1449,11 @@ read write iosize 512
     ${fd}   open libcurve file
     ${new_fd}       Convert To Integer    ${fd}
     ${rc}    write libcurve file error   ${new_fd}    length=512
-    ${expect_rc}    evaluate    int(-22)
-    should be equal    ${rc}   ${expect_rc}
-    ${rc}    read libcurve file   ${new_fd}   length=512
-    ${expect_rc}    evaluate    int(-22)
+    ${expect_rc}    evaluate    int(512)
```
[the remark this answers is not in the document]
Comment (wu-hanqing) [reply robot/curve_robot.txt:1452]: no, for read/write interface, when operation succeeded, it returned bytes that have read or written

File: src/client/request_closure.h [hunk 10]
```diff
@@ -149,6 +148,13 @@ class CURVE_CACHELINE_ALIGNMENT RequestClosure
...
+    RequestContext* reqCtx_ = nullptr;
+
+    // 当前request的tracker信息
```
[the remark this answers is not in the document]
Comment (wu-hanqing) [reply src/client/request_closure.h:155]: fix

File: src/client/request_closure.h [hunk 11]
```diff
@@ -178,6 +178,34 @@ class CURVE_CACHELINE_ALIGNMENT RequestClosure
...
+    }
+
+    RequestContext* AlignedRequest() const {
+        return alignedCtx_;
+    }
+
+ private:
+    void HandleRead();
+
+    void HandleWrite();
```
[the remark this answers is not in the document]
Comment (wu-hanqing) [reply src/client/request_closure.h:198]: fix

File: src/common/fast_align.h (hunk 2 above)
Comment (wu-hanqing) [reply src/common/fast_align.h:5]: fix

**Review by ilixiaocui — APPROVED** [review]

File: src/client/metacache.h [hunk 12]
```diff
@@ -228,6 +228,10 @@ class MetaCache {
         fileInfo_.filestatus = status;
     }
 
+    bool IsCloneFile() const {
+        return fileInfo_.filestatus == FileStatus::CloneMetaInstalled;
```
Comment (xu-chaojie) [inline src/client/metacache.h:232]: What abort FileStatus::Cloning ?

File: nbd/src/util.cpp [hunk 13]
```diff
@@ -173,6 +173,16 @@ int parse_args(std::vector<const char*>& args, std::ostream *err_msg,   // NOLIN
                 *err_msg << "curve-nbd: Invalid argument for sleep_ms!";
                 return -EINVAL;
             }
+        } else if (argparse_witharg(args, i, &cfg->block_size, err, "--block_size", (char*)(NULL))) {  // NOLINT
+            if (!err.str().empty()) {
+                *err_msg << "curve-nbd: " << err.str();
+                return -EINVAL;
+            }
+
+            if ((cfg->block_size % 512 != 0) && (cfg->block_size % 4096 != 0)) {
```
[the remark this answers is not in the document]
Comment (wu-hanqing) [reply nbd/src/util.cpp]: fix

File: src/client/metacache.h (hunk 12 above)
Comment (wu-hanqing) [reply src/client/metacache.h:232]: `FileStatus::Cloning` is unused now.

**Review by xu-chaojie — APPROVED** [review]

**Pull request merged**